tensor/indexed.go

```diff
@@ -0,0 +1,706 @@
+	}
+}
+
+// SortStableFunc stably sorts the row-wise indexes using given compare function.
+// The compare function operates directly on row numbers into the Tensor
+// as these row numbers have already been projected through the indexes.
+// cmp(a, b) should return a negative number when a < b, a positive
+// number when a > b and zero when a == b.
+// It is *essential* that it always returns 0 when the two are equal
+// for the stable function to actually work.
+func (ix *Indexed) SortStableFunc(cmp func(tsr Tensor, i, j int) int) {
+	ix.IndexesNeeded()
+	slices.SortStableFunc(ix.Indexes, func(a, b int) int {
+		return cmp(ix.Tensor, a, b) // key point: these are already indirected through indexes!!
+	})
+}
+
+// SortStable does stable default alpha or numeric sort.
+// Uses first cell of higher dimensional data.
+func (ix *Indexed) SortStable(ascending bool) {
+	if ix.Tensor.IsString() {
+		ix.SortStableFunc(func(tsr Tensor, i, j int) int {
+			return CompareAscending(tsr.StringRowCell(i, 0), tsr.StringRowCell(j, 0), ascending)
+		})
+	} else {
+		ix.SortStableFunc(func(tsr Tensor, i, j int) int {
+			return CompareAscending(tsr.FloatRowCell(i, 0), tsr.FloatRowCell(j, 0), ascending)
+		})
+	}
+}
+
+// FilterFunc is a function used for filtering that returns
+// true if Tensor row should be included in the current filtered
+// view of the tensor, and false if it should be removed.
+type FilterFunc func(tsr Tensor, row int) bool
+
+// Filter filters the indexes using given Filter function.
+// The Filter function operates directly on row numbers into the Tensor
+// as these row numbers have already been projected through the indexes.
+func (ix *Indexed) Filter(filterer func(tsr Tensor, row int) bool) {
+	ix.IndexesNeeded()
+	sz := len(ix.Indexes)
+	for i := sz - 1; i >= 0; i-- { // always go in reverse for filtering
+		if !filterer(ix.Tensor, ix.Indexes[i]) { // delete
+			ix.Indexes = append(ix.Indexes[:i], ix.Indexes[i+1:]...)
+		}
+	}
+}
+
+// Named arg values for FilterString
+const (
+	// Include means include matches
+	Include = false
+	// Exclude means exclude matches
+	Exclude = true
+	// Contains means the string only needs to contain the target string (see Equals)
+	Contains = true
+	// Equals means the string must equal the target string (see Contains)
+	Equals = false
+	// IgnoreCase means that differences in case are ignored in comparing strings
+	IgnoreCase = true
+	// UseCase means that case matters when comparing strings
+	UseCase = false
+)
+
+// FilterString filters the indexes using string values compared to given
+// string. Includes rows with matching values unless exclude is set.
+// If contains, only checks if row contains string; if ignoreCase, ignores case.
+// Use the named const args [Include], [Exclude], [Contains], [Equals],
+// [IgnoreCase], [UseCase] for greater clarity.
+// Uses first cell of higher dimensional data.
+func (ix *Indexed) FilterString(str string, exclude, contains, ignoreCase bool) { //types:add
```

Comment:
I would really prefer some sort of struct... Once you have a bunch of arguments with named bool constants, you should really just be using a struct.

Reply:
a9fb9164a struct